### PR TITLE
feat: execute markdown with jupytext frontmatter

### LINKED ISSUE #215: support for markdown with jupytext frontmatter?
Apologies if I've missed a discussion elsewhere. 
I see that at the moment, the mkdocs plugin `should_include` method checks whether files have extensions either an `ipynb` or `py` extension.  Would you consider adding support for markdown files (perhaps gated on whether they have a jupytext frontmatter)?   I can contribute, but wanted to check how you felt about it first

### PR DESCRIPTION
**Pull request by tlambert03**
closes #215.

@danielfrg, here is a simple implementation supporting markdown files.  As mentioned in #215, the biggest question for me is how best to cleanly add support for markdown files with jupytext frontmatter without causing any issues with "regular" markdown files that don't represent notebooks.

let me know if you have any thoughts

also, if you'd rather not check those two duplicated files for tests, we can probably create them as fixtures using jupytext to convert one of the other files

### CODE REVIEW

File: src/mkdocs_jupyter/plugin.py [hunk 1]
```diff
@@ -1,9 +1,10 @@
 import os
 import pathlib
 
+import jupytext
 import markdown
-from markdown.extensions.toc import TocExtension
 import mkdocs
+from markdown.extensions.toc import TocExtension
```
Comment (tlambert03) [inline src/mkdocs_jupyter/plugin.py:7]: i didn't do these btw... must have been from `just fmt`... are those not run during CI?

Comment (danielfrg) [reply src/mkdocs_jupyter/plugin.py:7]: oh weird. Maybe from your editor? No actions have run yet on this PR and the actions only report if something is broken they dont modify the code.

**Comment by danielfrg** [comment]
Yeah thats the one question I had how to differentiate files, and I like that you need to specify the kernel. that makes it explicit.

Can you add a test with a regular markdown file (no jupytext) too see what happens?
Is there like a hierarchy for which processor would take a file? I am not sure how mkdocs works when multiple plugins try to parse the same file.

File: src/mkdocs_jupyter/plugin.py (hunk 1 above)
Comment (tlambert03) [reply src/mkdocs_jupyter/plugin.py:7]: No no, sorry I _did_ run just fmt locally, I was just surprised that it was changing stuff that I didn't touch :)

Comment (danielfrg) [reply src/mkdocs_jupyter/plugin.py:7]: Ah ok, then yes. it came from that then :)

**Comment by tlambert03** [comment]
> Can you add a test with a regular markdown file (no jupytext) too see what happens?

there is already the index.md file in there ... and when I run `rye run mkdocs serve -f material-with-mds.yml` the logs don't try to render index.md:

```sh
➜ rye run mkdocs serve -f material-with-mds.yml
INFO    -  Building documentation...
INFO    -  Cleaning site directory
INFO    -  The following pages exist in the docs directory, but are not included in the "nav" configuration:
             - backquote_toc_test.ipynb
             - demo-script.py
             - demo.ipynb
             - fail.ipynb
             - ruby.ipynb
             - variational-inference-script.py
             - variational-inference.ipynb
             - extras/README.md
             - extras/styles.py
INFO    -  Converting notebook (execute=False):
           /Users/talley/dev/forks/mkdocs-jupyter/src/mkdocs_jupyter/tests/mkdocs/docs/backquote_toc_test.ipynb
INFO    -  Converting notebook (execute=True):
           /Users/talley/dev/forks/mkdocs-jupyter/src/mkdocs_jupyter/tests/mkdocs/docs/demo-jupytext.md
INFO    -  Converting notebook (execute=False):
           /Users/talley/dev/forks/mkdocs-jupyter/src/mkdocs_jupyter/tests/mkdocs/docs/demo-script.py
INFO    -  Converting notebook (execute=False):
           /Users/talley/dev/forks/mkdocs-jupyter/src/mkdocs_jupyter/tests/mkdocs/docs/demo.ipynb
INFO    -  Converting notebook (execute=False):
           /Users/talley/dev/forks/mkdocs-jupyter/src/mkdocs_jupyter/tests/mkdocs/docs/fail.ipynb
INFO    -  Converting notebook (execute=False):
           /Users/talley/dev/forks/mkdocs-jupyter/src/mkdocs_jupyter/tests/mkdocs/docs/ruby.ipynb
INFO    -  Converting notebook (execute=False):
           /Users/talley/dev/forks/mkdocs-jupyter/src/mkdocs_jupyter/tests/mkdocs/docs/variational-inference-script.py
INFO    -  Converting notebook (execute=False):
           /Users/talley/dev/forks/mkdocs-jupyter/src/mkdocs_jupyter/tests/mkdocs/docs/variational-inference.ipynb
INFO    -  Converting notebook (execute=True):
           /Users/talley/dev/forks/mkdocs-jupyter/src/mkdocs_jupyter/tests/mkdocs/docs/variational-inference.md
INFO    -  Converting notebook (execute=False):
           /Users/talley/dev/forks/mkdocs-jupyter/src/mkdocs_jupyter/tests/mkdocs/docs/extras/styles.py
```

here's what `index.md` looks like:

Screenshot 2024-08-27 at 9 28 10 PM

and `demo-jupytext.md`

Screenshot 2024-08-27 at 9 28 33 PM

anything you'd like me to explicitly test?

**Comment by tlambert03** [comment]
> Is there like a hierarchy for which processor would take a file? I am not sure how mkdocs works when multiple plugins try to parse the same file.

that's a good question that I don't have an immediate (knowledgeable) answer for.  I can say empirically that it seems to work fine with tests here and the repo i'm actually working on that motivated this PR... but i haven't yet looked yet for the exact line in mkdocs that would clarify *why* it works

**Comment by tlambert03** [comment]
> I will hold merging of until I read a bit more about how mkdocs handles that case so we know if this has any unexpected consequences 🙈 . I will try to do that soon.

yep, good idea.  I may have time today as well.  Will update if I learn more

**Comment by tlambert03** [comment]
ok, here's what I've learned:

as per the [mdocs Page Events docs](https://www.mkdocs.org/dev-guide/plugins/#page-events):

> The `pre_page` event is called before any actions are taken on the subject page and can be used to alter the `Page` instance.

(as I'm sure you know :))
so, from that alone, we can pretty much infer that if `Plugin.should_include(page.file)` returns `False`, then the page will be returned as is.  Side, note, based on [this line in mkdocs](https://github.com/mkdocs/mkdocs/blob/0b22a52cc86750beee45b61b7118480012cc804d/mkdocs/plugins.py#L569C19-L571), you could also have your plugin's `on_pre_page` method simply return `None` if `not self.should_include(page.file)` ... but it should also be fine to return the incoming page as is.

with this PR, `should_include` will return `False` for all markdown files, unless they specifically have kernel info in their metada.  For example, this is what you'd get for the "regular" `index.md` without any frontmatter:

```python
>>> jupytext.read(page.file.abs_src_path).get('metadata')
{'jupytext': {'main_language': 'javascript', 'notebook_metadata_filter': '-all', 'cell_metadata_filter': '-all', 'text_representation': {'extension': '.md', 'format_name': 'markdown'}}}
```

there is no `kernelspec` key at all, so `should_include` will return False, and we're good.

**Comment by hodgestar** [comment]
@tlambert03 @danielfrg I did a short review and this is looking great. Many thanks to both of you. :D

**Comment by danielfrg** [comment]
Ok, this looks good to me then. THank you for the PR and the research!

**Comment by danielfrg** [comment]
I will try to get this out later today :)

**Comment by tlambert03** [comment]
Thanks a bunch @danielfrg

**Comment by hodgestar** [comment]
Thank you!

**Pull request merged**